src/exo/cursors.py

```diff
@@ -0,0 +1,599 @@
+        return self._translate(Gap, self._path, 1 - dist)
+
+    def after(self, dist=1) -> Gap:
+        return self._translate(Gap, self._path, dist)
+
+    def prev(self, dist=1) -> Node:
+        return self._translate(Node, self._path, -dist)
+
+    def next(self, dist=1) -> Node:
+        return self._translate(Node, self._path, dist)
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (children)
+    # ------------------------------------------------------------------------ #
+
+    def child(self, attr, i=None) -> Node:
+        _node = getattr(self.node(), attr)
+        if i is not None:
+            _node = _node[i]
+        elif isinstance(_node, list):
+            raise ValueError('must index into block attribute')
+        cur = Node(self._proc, self._path + [(attr, i)])
+        # noinspection PyPropertyAccess
+        # cached_property is settable, bug in static analysis
+        cur._node = weakref.ref(_node)
+        return cur
+
+    def children(self) -> Iterable[Node]:
+        n = self.node()
+        # Top-level proc
+        if isinstance(n, LoopIR.proc):
+            yield from self._children_from_attrs(n, 'body')
+        # Statements
+        elif isinstance(n, (LoopIR.Assign, LoopIR.Reduce)):
+            yield from self._children_from_attrs(n, 'idx', 'rhs')
+        elif isinstance(n, (LoopIR.WriteConfig, LoopIR.WindowStmt)):
+            yield from self._children_from_attrs(n, 'rhs')
+        elif isinstance(n, (LoopIR.Pass, LoopIR.Alloc, LoopIR.Free)):
+            yield from []
+        elif isinstance(n, LoopIR.If):
+            yield from self._children_from_attrs(n, 'cond', 'body', 'orelse')
+        elif isinstance(n, (LoopIR.ForAll, LoopIR.Seq)):
+            yield from self._children_from_attrs(n, 'hi', 'body')
+        elif isinstance(n, LoopIR.Call):
+            yield from self._children_from_attrs(n, 'args')
+        # Expressions
+        elif isinstance(n, LoopIR.Read):
+            yield from self._children_from_attrs(n, 'idx')
+        elif isinstance(n, (LoopIR.Const, LoopIR.WindowExpr, LoopIR.StrideExpr,
+                            LoopIR.ReadConfig)):
+            yield from []
+        elif isinstance(n, LoopIR.USub):
+            yield from self._children_from_attrs(n, 'arg')
+        elif isinstance(n, LoopIR.BinOp):
+            yield from self._children_from_attrs(n, 'lhs', 'rhs')
+        elif isinstance(n, LoopIR.BuiltIn):
+            yield from self._children_from_attrs(n, 'args')
+
+    def _children_from_attrs(self, n, *args) -> Iterable[Node]:
+        for attr in args:
+            children = getattr(n, attr)
+            if isinstance(children, list):
+                for i in range(len(children)):
+                    yield self.child(attr, i)
+            else:
+                yield self.child(attr, None)
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (block selectors)
```

Comment:
What were our plans on statements/expression introspection?
I thought there were two ways:
1. Implement introspection function for all cases in cursors object e.g., `.is_if()` `.is_loop()`
2. Implement QAST-ish class object exposed to users so that they can write `if isinstance(QAST.Seq, s)`

I guess we are going with the 1?

Reply:
Not sure, we'll have to figure that out in a future PR. Right now I'm just looking at the `LoopIR` node directly via `isinstance` and `Node.node()`. I would be fine attaching introspection functions to cursors as in (1), to insulate the users from the underlying types. The `node()` function will have to be made private.

Reply:
+1 for figuring out the details when we need this in the future